hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/ArchivedTimelineWriter.java

```diff
@@ -0,0 +1,382 @@
+    updateVersionFile(newVersion);
+  }
+
+  private void updateVersionFile(int newVersion) throws IOException {
+    byte[] content = (String.valueOf(newVersion)).getBytes(StandardCharsets.UTF_8);
+    final Path versionFilePath = HoodieArchivedTimeline.getVersionFilePath(metaClient);
+    metaClient.getFs().delete(versionFilePath, false);
+    metaClient.getFs().createImmutableFileInPath(versionFilePath, Option.of(content));
+  }
+
+  /**
+   * Compacts the small archive files.
+   *
+   * <p>The parquet naming convention is:
+   *
+   * <pre>${min_instant}_${max_instant}_${level}.parquet</pre>
+   *
+   * <p>The 'min_instant' and 'max_instant' represent the instant time range of the parquet file.
+   * The 'level' represents the number of the level where the file is located, currently we
+   * have no limit for the number of layers.
+   *
+   * <p>These archive parquet files composite as an LSM tree layout, one parquet file contains
+   * a consecutive timestamp instant metadata entries. Different parquet files may have
+   * overlapping with the instant time ranges.
+   *
+   * <pre>
+   *   t1_t2_0.parquet, t3_t4_0.parquet, ... t5_t6_0.parquet       L0 layer
+   *                          \            /
+   *                             \     /
+   *                                |
+   *                                V
+   *                          t3_t6_1.parquet                      L1 layer
+   * </pre>
+   *
+   * <p>Compaction and cleaning: once the files number exceed a threshold(now constant 10) N,
+   * the oldest N files are then replaced with a compacted file in the next layer.
+   * A cleaning action is triggered right after the compaction.
+   *
+   * @param context HoodieEngineContext
+   */
+  @VisibleForTesting
+  public void compactAndClean(HoodieEngineContext context) throws IOException {
+    // 1. List all the latest snapshot files
+    HoodieArchivedManifest latestManifest = HoodieArchivedTimeline.latestSnapshotManifest(metaClient);
+    int layer = 0;
+    // 2. triggers the compaction for L0
+    Option<String> compactedFileName = doCompact(latestManifest, layer);
+    while (compactedFileName.isPresent()) {
+      // 3. once a compaction had been executed for the current layer,
+      // continues to trigger compaction for the next layer.
+      latestManifest.addFile(getFileEntry(compactedFileName.get()));
+      compactedFileName = doCompact(latestManifest, ++layer);
+    }
+
+    // cleaning
+    clean(context, layer);
+  }
+
+  private Option<String> doCompact(HoodieArchivedManifest manifest, int layer) throws IOException {
+    // 1. list all the files that belong to current layer
+    List<HoodieArchivedManifest.FileEntry> files = manifest.getFiles()
+        .stream().filter(file -> HoodieArchivedTimeline.isFileFromLayer(file.getFileName(), layer)).collect(Collectors.toList());
+
+    int compactionBatchSize = config.getArchiveMergeFilesBatchSize();
+
+    if (files.size() >= compactionBatchSize) {
```

Comment:
this needs to be total bytes or entries based?

Reply:
First entries based but also limited by the size.